### PR TITLE
Fix KeyError when using `cylc remove` after a reload changed the graph

### PR DESCRIPTION
**Pull request by MetRonnie**
Closes https://github.com/cylc/cylc-flow/issues/6502

Unreleased bug. To reproduce:
1. Start a workflow `a & a:started => b` where `a` is either a long running task or is set to fail
2. Change the graph to just `b` and reload the workflow
3. Do `cylc remove a`
4. This results in a `KeyError` which crashes the scheduler

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] No dependency changes 
- [x] Tests are included 
- [x] No changelog entry needed as unreleased bug
- [x] No docs needed
- [x] Master branch is correct

### CODE REVIEW

File: cylc/flow/scheduler.py [hunk 1]
```diff
@@ -1102,17 +1099,20 @@ def remove_tasks(
...
         if flow_nums is None:
             flow_nums = set()
-        # Mapping of *relative* task IDs to removed flow numbers:
-        removed: Dict[Tokens, FlowNums] = {}
-        not_removed: Set[Tokens] = set()
+        # Mapping of relative task IDs to removed flow numbers:
+        removed: Dict[str, FlowNums] = {}
+        not_removed: Set[str] = set()
+        # All the matched tasks (will add applicable active tasks below):
+        matched_tasks = inactive.copy()
```
Comment (MetRonnie) [inline cylc/flow/scheduler.py:1106]: Ok so I went back to using string IDs... But don't need the yucky `.split('/')` anymore

File: cylc/flow/task_pool.py [hunk 2]
```diff
@@ -2327,7 +2327,7 @@ def filter_task_proxies(
         ids: Iterable[str],
         warn_no_active: bool = True,
         inactive: bool = False,
-    ) -> 'Tuple[List[TaskProxy], Set[Tuple[str, PointBase]], List[str]]':
+    ) -> 'Tuple[List[TaskProxy], Set[Tuple[TaskDef, PointBase]], List[str]]':
```
Comment (MetRonnie) [inline cylc/flow/task_pool.py:2330]: One thing to note with this change is that the hashability/equality of `TaskDef`s is purely reference based (i.e. `__hash__()` and `__eq__()` fall back to `object` default). This should be fine as the `TaskDef`s are always coming from the `scheduler.config.taskdefs` dictionary, I believe

**Review by wxtim — APPROVED** [review]
:)

- [x] Reproduced bug
- [x] Checked that it fixed bug
- [x] Read code
- [x] Read tests
- [x] Got irked that "read" is pronounced differently dependent on tense without changing spelling

**Comment by hjoliver** [comment]
>  Got irked that "read" is pronounced differently dependent on tense without changing spelling

🤣  fair play


@dwsutherland could this be the reload bug your reported in our meeting the other day ?

**Review by hjoliver — APPROVED** [review]

**Pull request merged**